### PR TITLE
intel ssp dai

### PR DESCRIPTION
**Pull request by juimonen**
first attempt to port sof ssp into zephyr with dai interface
- this has the dai interface commit as base as it is not merged
- device tree definitions/integration needs fixing still
  - interrupts are now just "sof way" = string for level definition and running integer -> needs to be adapted to zephyr model
  - mn divider base address is now duplicated in all ssp nodes
- has lot of small helper functions that I'm not sure are somewhere available in zephyr, like io_reg_write, timer_poll_wait etc.

### CODE REVIEW

File: dts/xtensa/intel/intel_cavs25.dtsi
Comment (juimonen) [reply dts/xtensa/intel/intel_cavs25.dtsi:52]: yes will put in separate commit, but these clocks are used by mn divider code in ssp

File: drivers/dai/sof/ssp/ssp.h
Comment (dbaluta) [inline drivers/dai/sof/ssp/ssp.h:3]: why do we need to have a path with 'sof'. Shouldn't this driver be generic to Zephyr?

File: drivers/dai/sof/ssp/ssp.c
Comment (dbaluta) [inline drivers/dai/sof/ssp/ssp.c:2]: Any idea if we could reuse existing Zephyr I2S API? It looks there is a directory called drivers/i2s/

File: drivers/dai/sof/ssp/ssp.h
Comment (juimonen) [reply drivers/dai/sof/ssp/ssp.h:3]: yeah good question. I was starting to add dmic driver and flat directory was becoming really messy. I tried to browse the zephyr driver tree and almost all directories are flat and the drivers are code wise smaller than this. Some of the drivers I'm trying to port are quite big and have auxiliary headers. I could try to lump all of that into 1 .c and 1 .h but that would make those files huge.

So yes, this should be (and is) generic driver, just I want to make it clear what is part of what when more drivers are appearing. Maybe remove sof and just add "ssp" dir?

File: drivers/dai/sof/ssp/ssp.c
Comment (juimonen) [reply drivers/dai/sof/ssp/ssp.c:2]: Probably would be possible. But the dai API (that this ssp driver is using) is like 1 level up in generic terms, so the idea is that we can gather many other audio drivers, than just i2s, behind it. So the application using the API (like sof) can use the same API with different drivers, like dmic, hda, alh etc.

File: drivers/dai/sof/ssp/ssp.h
Comment (dbaluta) [reply drivers/dai/sof/ssp/ssp.h:3]: @juimonen I dont like flat directories, anyhow I was suggesting that `sof` in the name of the directory is not OK as we are trying to write application agnostic drivers.

My first preference for SSP driver should be something like: `drivers/i2s/intel/ssp.c`. If we decide that `i2s` interface is not what we want then we can go with `drivers/dai/intel/ssp.c`

Comment (juimonen) [reply drivers/dai/sof/ssp/ssp.h:3]: ack.

File: dts/xtensa/intel/intel_cavs25.dtsi
Comment (juimonen) [reply dts/xtensa/intel/intel_cavs25.dtsi:52]: now in separate commit

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: ok, changing to those.

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: changing to that, intial tests seemed to work

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: ok this is nice, changed to this

**Comment by dbaluta** [comment]
@juimonen what i think it is missing here is a proper framework that deals with DAIs in Zephyr. Can we schedule a meeting next week with al interesting parties. 

Zephyr misses a proper audio stack(?) and I think it is important to get the interfaces right from the beginning.

**Comment by juimonen** [comment]
> @juimonen what i think it is missing here is a proper framework that deals with DAIs in Zephyr. Can we schedule a meeting next week with al interesting parties.

@dbaluta yes sure let's have a meeting. @lgirdwood FYI.

**Comment by kv2019i** [comment]
@dbaluta wrote:
> Zephyr misses a proper audio stack(?) and I think it is important to get the interfaces right from the beginning.

That's what this RFC is really all about.  The DAI interface here is aligned with SOF and has been used and supported by multiple vendors (in SOF), so it would seem like a reasonable start for this work. This DAI interface should be useful for other users in Zephyr and certainly SOF could use it (and we could move hw drivers over to Zephyr). We could of course shoot for higher level integration, but that may be a much longer road as different users might have more varied reqs.

The existing I2S interface doesn't really cut it for SOF, as SOF needs to support other interfaces (already now). But with this DAI interface, we could share a considerable chunk of code with Zephyr and benefit both projects. That's not to say we shouldn't/couldn't aim higher, but at least that's a concrete way to start.

**Comment by marcinszkudlinski** [comment]
I have doubts if we need to have entire DAI as a driver that have all HW releated code inside - like keeping all code under drivers/dai/ssp/*
I understand the idea of DAI abstraction, yet I think it should be used as an abstracton, creating a common interface for drivers located in - in case of SSP - drivers/i2s/i2s_cavs, keeping shimm / config layer in dai subdir

It may not be possible / practical in every situation, however it is worth to think about. Otherwise we'll have a huge code duplication in drivers

**Comment by lgirdwood** [comment]
> > @juimonen what i think it is missing here is a proper framework that deals with DAIs in Zephyr. Can we schedule a meeting next week with al interesting parties.
> 
> @dbaluta yes sure let's have a meeting. @lgirdwood FYI.

@dbaluta let me know what suits if you need a call - intention here is to incrementally add DAI features to align with a similar feature set and abstraction to SOF  and Linux DAI interfaces. I think you are probably thinking of the layers that go on top of this bottom level e.g. APIs for connecting DMAs to DAI, or defining an audio endpoint and these come later.

File: CODEOWNERS
Comment (lgirdwood) [inline CODEOWNERS:245]: What about Arsen and  @dbaluta here too ?

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h
Comment (lgirdwood) [inline drivers/dai/intel/ssp/dai-params-intel-ipc4.h:17]: This needs to be prefixed to align with the owner/usage. e.g.
```dai_intel_ipc4_...``` for all enums, structs here. This ways it's obvious its for Intel IPC4 only users

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h
Comment (lgirdwood) [inline drivers/dai/intel/ssp/dai-params-intel-ipc4.h:12]: This could collide with other I2S TDM headers/users. Can you change prefix to ```DAI_INTEL_TDM_...```

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h
Comment (lgirdwood) [inline drivers/dai/intel/ssp/dai-params-intel-ipc4.h:73]: ditto prefix changes for all structs, enum and macros

File: drivers/dai/intel/ssp/dai-params-intel.h
Comment (lgirdwood) [inline drivers/dai/intel/ssp/dai-params-intel.h:7]: IPC3 needs to be added here.

File: drivers/dai/intel/ssp/dai-params-intel.h
Comment (lgirdwood) [inline drivers/dai/intel/ssp/dai-params-intel.h:13]: ```DAI_INTEL_IPC3_...``` . Similar prefix changes need to be made here in this file too.

File: drivers/dai/intel/ssp/ssp.c
Comment (lgirdwood) [inline drivers/dai/intel/ssp/ssp.c:35]: This should be define in either the IPC3/4 header as it's configured externally by clients ?

**Comment by dbaluta** [comment]
@lgirdwood i'm fine with this for now! we should schedule a TSC meeting this month.

File: drivers/dai/intel/ssp/ssp.h
Comment (lgirdwood) [inline drivers/dai/intel/ssp/ssp.h:13]: Lots of macros here that could collide. Can we namespace with ```SSP_INTEL_``` and reuse similar macros from Zephyr common headers. e.g. there are probably macros to set bits etc ?

File: drivers/dai/intel/ssp/ssp.h
Comment (lgirdwood) [inline drivers/dai/intel/ssp/ssp.h:349]: This is also quite a common name we would need to namespace with a common prefix here.

File: drivers/dai/intel/ssp/ssp.c
Comment (lgirdwood) [inline drivers/dai/intel/ssp/ssp.c:741]: We need a timeout here so we never spin forever.

File: drivers/dai/intel/ssp/ssp.c
Comment (lgirdwood) [inline drivers/dai/intel/ssp/ssp.c:759]: timeout.

**Review by lgirdwood — DISMISSED** [review]
Mostly fine, however we do need to 
1) namespace a lot of the Intel SSP macros, structs etc so we never collide with other users
2) Never spin forever on HW bit changes without a timeout
3) I would also use the IPC3/4 KConfig to reduce image size and only build what'w needed.
4) reuse any common macros, e.g. I suspect Zephyr has a ```SET_BIT()``` that should be used here etc.

**EDIT** Fwiw, the DAI API part is good to go, we just need more work on the SSP part.

File: CODEOWNERS
Comment (juimonen) [reply CODEOWNERS:245]: yes, please just suggest the correct persons here, I just silenced the CI with my own username

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h
Comment (juimonen) [reply drivers/dai/intel/ssp/dai-params-intel-ipc4.h:12]: not needed, removed.

File: drivers/dai/intel/ssp/ssp.h
Comment (marcinszkudlinski) [inline drivers/dai/intel/ssp/ssp.h:373]: struct dai is a generic structure from sof-dai, should be used for all dai-based drivers, not only for ssp. So pls move it out of ssp.h
same applies to struct dai_plat_data and all sub structures

File: drivers/dai/intel/ssp/ssp.c
Comment (juimonen) [reply drivers/dai/intel/ssp/ssp.c:35]: removed

File: drivers/dai/intel/ssp/ssp.h
Comment (juimonen) [reply drivers/dai/intel/ssp/ssp.h:349]: ok

File: drivers/dai/intel/ssp/ssp.h
Comment (juimonen) [reply drivers/dai/intel/ssp/ssp.h:13]: yes will prefix. I only found BIT() from zephyr...

File: drivers/dai/intel/ssp/ssp.c
Comment (juimonen) [reply drivers/dai/intel/ssp/ssp.c:741]: ok

File: drivers/dai/intel/ssp/ssp.c
Comment (juimonen) [reply drivers/dai/intel/ssp/ssp.c:759]: ok

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h
Comment (juimonen) [reply drivers/dai/intel/ssp/dai-params-intel-ipc4.h]: ok

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h
Comment (juimonen) [reply drivers/dai/intel/ssp/dai-params-intel-ipc4.h]: ok

File: drivers/dai/intel/ssp/dai-params-intel.h
Comment (juimonen) [reply drivers/dai/intel/ssp/dai-params-intel.h]: ok

File: drivers/dai/intel/ssp/dai-params-intel.h
Comment (juimonen) [reply drivers/dai/intel/ssp/dai-params-intel.h]: ok

File: drivers/dai/intel/ssp/ssp.c
Comment (marcinszkudlinski) [inline drivers/dai/intel/ssp/ssp.c:52]: Looks like there are many functions common for all DAI-based drivers. Can we create a library or a helper instead of repeating dai_ssp_update_bit, dai_dmic_update_bit, dai_XYZ_update_bit ?

File: drivers/dai/dai-params-intel-ipc4.h
Comment (juimonen) [reply drivers/dai/dai-params-intel-ipc4.h]: these are now separated into 2 different files, for 2 different versions of the bespoke data

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: ok rebased, as it is merged

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: ok, I tried to trace the origins of the negative conversion in sof, without too much luck. I also tried to see if it is called with possibly negative values somewhere, but did't find any candidates. So I would leave it as it is now, and refactor this later, possibly move to zephyr math etc.

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: is this really so? I didn't now get any warnings from CI build...

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: ok, see above..

File: drivers/dai/intel/ssp/ssp.c
Comment (juimonen) [reply drivers/dai/intel/ssp/ssp.c:52]: yes there's overlap, but there are not too many. I did this when creating blob generator for alsa-utils. I would actually like to move the common part to zephyr side.

If you look at other zephyr drivers, almost all define they own bit macros, so we are not very badly aligned in that sense. So I would refactor this then, when I bring in other drivers than ssp. As said the overlap is not big and I will probably not get it right without other driver -> anyway I would need to revisit this.

File: drivers/dai/ssp.c
Comment (teburd) [reply drivers/dai/ssp.c]: yep, first bullet here https://docs.zephyrproject.org/latest/contribute/guidelines.html?highlight=styling#coding-style

Comment (juimonen) [reply drivers/dai/ssp.c]: ok thanks, yes maybe I could read that... :) anyway, will change

File: drivers/dai/intel/ssp/ssp.c
Comment (juimonen) [reply drivers/dai/intel/ssp/ssp.c:34]: I removed now many of the extra set bits etc and used zephyr sys_read and write

File: drivers/dai/ssp.c
Comment (juimonen) [reply drivers/dai/ssp.c]: changed now

File: doc/hardware/peripherals/audio/dai.rst
Comment (kv2019i) [inline doc/hardware/peripherals/audio/dai.rst:12]: Nice to have updates: mention the history of the interface in SOF project

**Review by kv2019i — APPROVED** [review]
Excellent, looks good now @juimonen ! I especially like the test case, this is a great thing to have and something we can build on later.

File: drivers/dai/intel/ssp/ssp.c
Comment (lyakh) [inline drivers/dai/intel/ssp/ssp.c:1522]: could somebody, please, split this function... later

File: drivers/dai/intel/ssp/ssp.c
Comment (lyakh) [inline drivers/dai/intel/ssp/ssp.c:1874]: this is in fact interesting. I assume you have to increase `CONFIG_HEAP_MEM_POOL_SIZE` for this to work. Then - is the returned address cached or uncached? We use uncached so far. It seems to me, that this will be allocated from the `.noinit` section and it will be uncached too

File: tests/boards/intel_adsp/ssp/Kconfig
Comment (lyakh) [inline tests/boards/intel_adsp/ssp/Kconfig:11]: wondering, what happens if this isn't selected? Can SSPs still be used properly with all the sampling rates?

**Review by lyakh — APPROVED** [review]
just a couple of questions and possible improvement suggestions, that can be addressed after the initial merge

**Comment by ArsenEloglian** [comment]
...PLEASE DON'T MERGE THIS PR...

Please review this one:
https://github.com/zephyrproject-rtos/zephyr/pull/44870

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]
Please DON'T merge this PR, it contains way to much code that is not yet ready.
Proceed with the one Arsen created, it contains a subset of changes extracted from this PR

**Review by marcinszkudlinski — APPROVED** [review]
After couple of additional discussions - please proceed with merging

**Review by lgirdwood — APPROVED** [review]
quick review during vacation, all my review comments addressed.

**Pull request merged**
